tools/harness-automation/parse_topofile.py

```
@@ -0,0 +1,123 @@
+                    break
+                if re.match(r'#.*', line):
+                    continue
+
+                matched_case = re.match(r'(.*)-(.*)', line, re.M | re.I)
+
+                if case != 'all' and case != matched_case.group(1):
+                    continue
+
+                logging.info('case %s:' % matched_case.group(1))
+                case_count += 1
+                role_vendor_str = matched_case.group(2)
+                role_vendor_raw_list = re.split(',', role_vendor_str)
+                role_vendor_list = []
+
+                for i in range(len(role_vendor_raw_list)):
+                    device_pair = re.split(':', role_vendor_raw_list[i])
+                    role_vendor_list.append(tuple(device_pair))
+                logging.info('\trole-vendor pair: %s' % role_vendor_list)
+
+                for i in range(len(role_vendor_list)):
+                    if role_vendor_list[i][1] in case_vendor_dict:
+                        case_vendor_dict[role_vendor_list[i][1]] += 1
+                    else:
+                        case_vendor_dict[role_vendor_list[i][1]] = 1
+
+                    if role_vendor_list[i][1] in testbed_vendor_dict:
+                        if testbed_vendor_dict[role_vendor_list[i][1]] < case_vendor_dict[role_vendor_list[i][1]]:
+                            testbed_vendor_dict[role_vendor_list[i][1]] = case_vendor_dict[role_vendor_list[i][1]]
+                    else:
+                        testbed_vendor_dict[role_vendor_list[i][1]] = 1
+
+                logging.info('\tneeded vendor devices:%s' % case_vendor_dict)
+            f.close
+
+        count_any = MAX_VENDOR_DEVICE
+        for key in testbed_vendor_dict:
+            if key != 'Any':
+                count_any -= testbed_vendor_dict[key]
+        if 'Any' in testbed_vendor_dict:
+            testbed_vendor_dict['Any'] = count_any
+
+        logging.info('\nTestbed needed vendor devices:%s' % testbed_vendor_dict)
+    except Exception as e:
```

just let exception be raised, rather than catch it?